From da32c7dffc49f1a8c75e98f0a9a843e5d22bb497 Mon Sep 17 00:00:00 2001
From: Marc Mutz <marc.mutz@qt.io>
Date: Tue, 21 Oct 2025 18:10:31 +0200
Subject: [PATCH 542/553] QUnicodeTools: port getWordBreaks() to
 QStringIterator

This one is a bit more complicated than the previous two, since
there's a nested loop. Solve it by using a copy of the QStringIterator
for look-ahead loop.

To see that old and new version are equivalent, observe that qsizetype
`i` and `lookahead` always pointed _onto_, while QStringIterator
always points to just _after_ the last-consumed code-unit.

Pick-to: 6.8 6.5
Change-Id: I391fafcf2418dac39b4aea3b3a3a675114233dff
Reviewed-by: Edward Welbourne <edward.welbourne@qt.io>
(cherry picked from commit 9e442a8e27000c7b925aa54a5a4c59e5ca2badb7)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
---
 src/corelib/text/qunicodetools.cpp | 27 +++++++--------------------
 1 file changed, 7 insertions(+), 20 deletions(-)

diff --git a/src/corelib/text/qunicodetools.cpp b/src/corelib/text/qunicodetools.cpp
index a1867f747f9..56fa41c51ab 100644
--- a/src/corelib/text/qunicodetools.cpp
+++ b/src/corelib/text/qunicodetools.cpp
@@ -242,16 +242,10 @@ static void getWordBreaks(const char16_t *string, qsizetype len, QCharAttributes
     QUnicodeTables::WordBreakClass cls = QUnicodeTables::WordBreak_LF; // to meet WB1
     auto real_cls = cls; // Unaffected by WB4
 
-    for (qsizetype i = 0; i != len; ++i) {
-        const qsizetype pos = i;
-        char32_t ucs4 = string[i];
-        if (QChar::isHighSurrogate(ucs4) && i + 1 != len) {
-            ushort low = string[i + 1];
-            if (QChar::isLowSurrogate(low)) {
-                ucs4 = QChar::surrogateToUcs4(ucs4, low);
-                ++i;
-            }
-        }
+    QStringIterator it(QStringView{string, len});
+    while (it.hasNext()) {
+        const qsizetype pos = it.index();
+        const char32_t ucs4 = it.nextOrRawCodeUnit();
 
         const auto prop = QUnicodeTables::properties(ucs4);
         QUnicodeTables::WordBreakClass ncls = (QUnicodeTables::WordBreakClass) prop->wordBreakClass;
@@ -294,15 +288,8 @@ static void getWordBreaks(const char16_t *string, qsizetype len, QCharAttributes
             break;
         case WB::Lookup:
         case WB::LookupW:
-            for (qsizetype lookahead = i + 1; lookahead < len; ++lookahead) {
-                char32_t ucs4 = string[lookahead];
-                if (QChar::isHighSurrogate(ucs4) && lookahead + 1 != len) {
-                    ushort low = string[lookahead + 1];
-                    if (QChar::isLowSurrogate(low)) {
-                        ucs4 = QChar::surrogateToUcs4(ucs4, low);
-                        ++lookahead;
-                    }
-                }
+            for (auto lookahead = it; lookahead.hasNext(); /**/) {
+                const char32_t ucs4 = lookahead.nextOrRawCodeUnit();
 
                 const auto prop = QUnicodeTables::properties(ucs4);
                 QUnicodeTables::WordBreakClass tcls = (QUnicodeTables::WordBreakClass) prop->wordBreakClass;
@@ -314,7 +301,7 @@ static void getWordBreaks(const char16_t *string, qsizetype len, QCharAttributes
 
                 if (Q_LIKELY(tcls == cls || (action == WB::LookupW && (tcls == QUnicodeTables::WordBreak_HebrewLetter
                                                                        || tcls == QUnicodeTables::WordBreak_ALetter)))) {
-                    i = lookahead;
+                    it = lookahead;
                     ncls = tcls;
                     action = WB::NoBreak;
                 }
-- 
2.51.2

